File: src/main/LoginSystem.java
package main;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

import DataBaseController.AdvisorDBController;
import DataBaseController.DepartmentSchedulerDBController;
import DataBaseController.StudentDBController;
import com.fasterxml.jackson.databind.ObjectMapper;

public class LoginSystem {
    private static final String USERS_DIRECTORY = "CSE3063F24P1_GRP17/src/database/students"; // Kullanıcı dosyalarının olduğu dizin

    // These Student, Advisor and Department Scheduler objects down below will hold the loaded objects that represents the current user that
    // is successfully logged in to the system

    // Also they have their own DBControllers for loading their information from the database individually
    private Student student;
    private StudentDBController studentDBController;

    private Advisor advisor;
    private AdvisorDBController advisorDBController;

    private DepartmentScheduler departmentScheduler;
    private DepartmentSchedulerDBController departmentSchedulerDBController;

    LoginSystem(StudentDBController studentDBController,AdvisorDBController advisorDBController,DepartmentSchedulerDBController departmentSchedulerDBController) {
        this.studentDBController = studentDBController;
        this.advisorDBController = advisorDBController;
        this.departmentSchedulerDBController = departmentSchedulerDBController;
    }


    /**
     * Authenticates a user (either a student or an advisor) based on their nickname and password.
     *
     * @param studentId The user's nickname (ID).
     * @param password The user's password.
     * @return true if authentication is successful, false otherwise.
     * @throws IOException if an I/O error occurs.
     */
    // authenticateUser is responsible of the correctness of the login information given by the student user and matches it's information
    // from the database and loads the necessary objects related to the student user
    public boolean authenticateUser(String studentId, String password) throws IOException {
        // Try to load the student from the json files
        // With the given student ID (also means the nickname), we look for that particular student in our database that has the same file name
        if(studentDBController.loadStudent(studentId)) {
            this.student = studentDBController.getStudent();
        }
        else{
            return false;
        }

        // Student has been loaded and it's time for checking the password
        if (student.getPassword().equals(password)) {
            return true;
        }
        // Password checking has been failed, so delete the student object and return null to them
        // Also return false to indicate the error in login attempt
        else{
            this.student = null;
            studentDBController.setStudent(null);
            return false;
        }
    }
    // authenticateUser is responsible of the correctness of the login information given by the advisor user and matches it's information
    // from the database and loads the necessary objects related to the advisor user
    public boolean authenticateAdvisorUser(String advisorId, String password) throws IOException{
        if(advisorDBController.loadAdvisor(advisorId)){
            this.advisor = advisorDBController.getAdvisor();
        }else{
            return false;
        }

        if(advisor.getPassword().equals(password)){
            return true;
        }else{
            this.advisor = null;
            advisorDBController.setAdvisor(null);
            return false;
        }

    }
    // authenticateUser is responsible of the correctness of the login information given by the department scheduler user and matches it's information
    // from the database and loads the necessary objects related to the department scheduler user
    public boolean authenticateDepartmentSchedulerUser(String departmentSchedulerId, String password) throws IOException{

        if(departmentSchedulerDBController.loadDepartmentScheduler(departmentSchedulerId)){
            this.departmentScheduler = departmentSchedulerDBController.getDepartmentScheduler();
        }else{
            return false;
        }

        if(departmentScheduler.getPassword().equals(password)){
            return true;
        }else{
            this.departmentScheduler = null;
            departmentSchedulerDBController.setDepartmentScheduler(null);
            return false;
        }

    }

    // Getter methodu
    public Student getStudent() {
        return student;
    }

    public Advisor getAdvisor() {
        return advisor;
    }

    public DepartmentScheduler getDepartmentScheduler() {
        return departmentScheduler;
    }

    public void setDepartmentScheduler(DepartmentScheduler departmentScheduler) {
        this.departmentScheduler = departmentScheduler;
    }

}

File: main/LoginSystem.py
import sys
import os
# Üst dizinin yolunu ekle
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from databaseController.DepartmentSchedulerDBController import DepartmentSchedulerDBController
from main.DepartmentScheduler import DepartmentScheduler
from databaseController.AdvisorDBController import AdvisorDBController
from databaseController.StudentDBController import StudentDBController
from databaseController.AdminDBController import AdminDBController
from databaseController.DepartmentHeadDBController import DepartmentHeadDBController
from main.Advisor import Advisor
from main.Student import Student
from main.Admin import Admin
from main.DepartmentHead import DepartmentHead


class LoginSystem:

    __advisor: Advisor
    __student: Student
    __departmentScheduler: DepartmentScheduler
    __admin : Admin
    __departmentHead : DepartmentHead
    __advisorDBController: AdvisorDBController
    __studentDBController: StudentDBController
    __departmentSchedulerDBController: DepartmentSchedulerDBController
    __adminDBController: AdminDBController
    __departmentHeadDBController: DepartmentHeadDBController

    def __init__(self, studentDBController, advisorDBController, departmentSchedulerDBController, adminDBController, departmentHeadDBController):
        self.__studentDBController = studentDBController
        self.__advisorDBController = advisorDBController
        self.__departmentSchedulerDBController = departmentSchedulerDBController
        self.__adminDBController = adminDBController
        self.__departmentHeadDBController = departmentHeadDBController

    # Getter and setter for Advisor
    def getAdvisor(self):
        return self.__advisor

    def setAdvisor(self, advisor):
        self.__advisor = advisor

    # Getter and setter for Student
    def getStudent(self):
        return self.__student

    def setStudent(self, student):
        self.__student = student

    def getDepartmentScheduler(self):
        return self.__departmentScheduler

    def setDepartmentScheduler(self, departmentScheduler):
        self.__departmentScheduler = departmentScheduler

    def getAdmin(self):
        return self.__admin
    
    def setAdmin(self, admin):
        self.__admin = admin
    
    def getDepartmentHead(self):
        return self.__departmentHead
    
    def setDepartmentHead(self, departmentHead):
        self.__departmentHead = departmentHead

    # Getter and setter for AdvisorDBController
    def getAdvisorDBController(self):
        return self.__advisorDBController

    def setAdvisorDBController(self, advisorDBController):
        self.__advisorDBController = advisorDBController

    # Getter and setter for StudentDBController
    def getStudentDBController(self):
        return self.__studentDBController

    def setStudentDBController(self, studentDBController):
        self.__studentDBController = studentDBController

    def getAdminDBController(self):
        return self.__adminDBController
    
    def setAdminDBController(self, adminDBController):
        self.__adminDBController = adminDBController

    # Authenticate Advisor
    def authenticateAdvisorUser(self, advisorId, password):
        if self.__advisorDBController.loadAdvisor(advisorId):
            self.__advisor = self.__advisorDBController.getAdvisor()
        else:
            return False

        if self.__advisor.getPassword() == password:
            return True
        else:
            self.__advisor = None
            self.__advisorDBController.setAdvisor(None)
            return False

    # Authenticate Student
    def authenticateStudentUser(self, studentId, password):
        if self.__studentDBController.loadStudent(studentId):
            self.__student = self.__studentDBController.getStudent()
        else:
            return False

        if self.__student.getPassword() == password:
            return True
        else:
            self.__student = None
            self.__studentDBController.setStudent(None)
            return False

    # Authenticate Department Scheduler
    def authenticateDepartmentSchedulerUser(self, departmentSchedulerId, password):
        if self.__departmentSchedulerDBController.loadDepartmentScheduler(departmentSchedulerId):
            self.__departmentScheduler = self.__departmentSchedulerDBController.getDepartmentScheduler()
        else:
            return False

        if self.__departmentScheduler.getPassword() == password:
            return True
        else:
            self.__departmentScheduler = None
            self.__departmentSchedulerDBController.setDepartmentScheduler(None)
            return False
        
    def authenticateAdminUser(self, adminId, password):
        if self.__adminDBController.loadAdmin(adminId):
            self.__admin = self.__adminDBController.getAdmin()
        else:
            return False

        if self.__admin.getPassword() == password:
            return True
        else:
            self.__admin = None
            self.__adminDBController.setAdmin(None)
            return False
    
    def authenticateDepartmentHead(self, departmentHeadId, password):
        if self.__departmentHeadDBController.loadDepartmentHead(departmentHeadId):
            self.__departmentHead = self.__departmentHeadDBController.getDepartmentHead()
        else:
            return False

        if self.__departmentHead.getPassword() == password:
            return True
        else:
            self.__departmentHead = None
            self.__departmentHeadDBController.setDepartmentHead(None)
            return False
        